tuning/sft_trainer.py

```diff
@@ -344,34 +353,103 @@ def main(**kwargs):  # pylint: disable=unused-argument
         help='Pass a json string representing K:V pairs to be associated\
               to the tuning run in the tracker. e.g. \'{"gpu":"A100-80G"}\'',
     )
-    (
+    return parser
+
+
+def parse_arguments(parser, json_config=None):
+    # accept arguments via command-line or JSON
+    if json_config:
+        (
+            model_args,
+            data_args,
+            training_args,
+            trainer_controller_args,
+            lora_config,
+            prompt_tuning_config,
+            file_logger_config,
+            aim_config,
+        ) = parser.parse_dict(json_config, allow_extra_keys=True)
+        peft_method = json_config.get("peft_method")
+        exp_metadata = json_config.get("exp_metadata")
+    else:
+        (
+            model_args,
+            data_args,
+            training_args,
+            trainer_controller_args,
+            lora_config,
+            prompt_tuning_config,
+            file_logger_config,
+            aim_config,
+            additional,
+            _,
+        ) = parser.parse_args_into_dataclasses(return_remaining_strings=True)
+
+        peft_method = additional.peft_method
+        exp_metadata = additional.exp_metadata
+
+    if peft_method == "lora":
+        tune_config = lora_config
+    elif peft_method == "pt":
+        tune_config = prompt_tuning_config
+    else:
+        tune_config = None
+
+    return (
         model_args,
         data_args,
         training_args,
         trainer_controller_args,
-        lora_config,
-        prompt_tuning_config,
+        tune_config,
         file_logger_config,
         aim_config,
-        additional,
-        _,
-    ) = parser.parse_args_into_dataclasses(return_remaining_strings=True)
+        exp_metadata,
+    )
 
+
+def main(**kwargs):  # pylint: disable=unused-argument
     logger = logging.get_logger("__main__")
 
-    peft_method = additional.peft_method
-    if peft_method == "lora":
-        tune_config = lora_config
-    elif peft_method == "pt":
-        tune_config = prompt_tuning_config
-    else:
-        tune_config = None
+    parser = get_parser()
+    job_config = get_json_config()
```

Comment:
We don't want to move it inside the function because we want to be able to call sft_trainer with either the JSON config or command-line arguments. Many users of the library don't use JSON to call SFTTrainer. But if we are running within the image (aka with `accelerate_launch.py`) then we expect a JSON to be passed and thus the error.